src/app/course/course-list/course-list.component.ts

```diff
@@ -1,49 +1,69 @@
-import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
+import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
 import { Router } from '@angular/router';
+import { Subject, Observable } from 'rxjs';
+import {
+  debounceTime, distinctUntilChanged, map, switchMap, tap
+} from 'rxjs/operators';
 import { Course } from 'src/app/models/data-model';
 import { FilterPipe } from 'src/app/shared/pipes/filter.pipe';
 import { CourseInstance } from '../course';
 import { CourseService } from '../course.service';
+import { LoaderService } from 'src/app/services/loader.service';
 
 @Component({
   selector: 'app-course-list',
   templateUrl: './course-list.component.html',
   styleUrls: ['./course-list.component.scss'],
   providers: [FilterPipe],
-  //changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class CourseListComponent implements OnInit {
-  @Input() courseList: Course[];
   @Output()
   deleteCourseItem: EventEmitter<number> = new EventEmitter<number>();
 
   public message: string;
   public listBySearch: Course[];
+  
+  courseList: Course[];
+
+  public filterSubject= new Subject<string>();
+  public query: string;
 
   
   constructor(
     private courseService: CourseService,
     private router: Router,
-    // private cdRef :ChangeDetectorRef,
-    ) {}
+    ) {
+      this.filterSubject.pipe(switchMap(data => this.courseService.searchCourse(data)

```

Comment:
It doesn't work as expected. Check first task in homework. The <input> is supposed to emit a value whenever user presses a key. Then `debounce` and extra checks should be applied. Only after that the request to API should be made